satellite/satellitedb/containment.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/audit"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type containment struct {
+	db *dbx.DB
+}
+
+// Get gets the pending audit by node id
+func (containment *containment) Get(ctx context.Context, id pb.NodeID) (*audit.PendingAudit, error) {
+	if id.IsZero() {
+		return nil, audit.ContainError.New("node ID empty")
+	}
+
+	pending, err := containment.db.Get_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(id.Bytes()))
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, audit.ErrContainedNotFound.New(id.String())
+		}
+		return nil, audit.ContainError.Wrap(err)
+	}
+
+	return convertDBPending(pending)
+}
+
+// IncrementPending creates a new pending audit entry, or increases its reverify count if it already exists
+func (containment *containment) IncrementPending(ctx context.Context, pendingAudit *audit.PendingAudit) error {
+	tx, err := containment.db.Open(ctx)
+	if err != nil {
+		return audit.ContainError.Wrap(err)
+	}
+
+	existingAudit, err := tx.Get_PendingAudits_By_NodeId(ctx, dbx.PendingAudits_NodeId(pendingAudit.NodeID.Bytes()))
+	if err == sql.ErrNoRows {
+		statement := containment.db.Rebind(
+			`INSERT INTO pending_audits (node_id, piece_id, stripe_index, share_size, expected_share_hash, reverify_count)
+			VALUES (?, ?, ?, ?, ?, ?)`,
+		)
+		_, err = tx.Tx.ExecContext(ctx, statement,
+			pendingAudit.NodeID.Bytes(), pendingAudit.PieceID.Bytes(), pendingAudit.StripeIndex, pendingAudit.ShareSize, pendingAudit.ExpectedShareHash, pendingAudit.ReverifyCount,
+		)
+		if err != nil {
+			return audit.ContainError.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+	} else if err == nil {
+		if !bytes.Equal(existingAudit.ExpectedShareHash, pendingAudit.ExpectedShareHash) {
+			return audit.ContainError.Wrap(errs.Combine(Error.New("pending audit already exists for nodeID %s", pendingAudit.NodeID), tx.Rollback()))
```

Comment:
Should the caller distinguish this error from the others for taking a different action?

Reply:
Revolved in a conversation in other means.